numba/targets/linalg.py

```diff
@@ -2425,3 +2425,71 @@ def _1d_matrix_rank_impl(a, tol):
             assert 0  # unreachable
 
     return _get_matrix_rank_impl(a, tol)
+
+
+@overload(np.linalg.matrix_power)
+def matrix_power_impl(a, n):
+    """
+    Computes matrix power. Only integer powers are supported in numpy.
+    """
+
+    _check_linalg_matrix(a, "matrix_power")
+    np_dtype = np_support.as_dtype(a.dtype)
+
+    nt = getattr(n, 'dtype', n)
+    if not isinstance(nt, types.Integer):
+        raise TypeError("Exponent must be an integer.")
+
+    def matrix_power_impl(a, n):
+
+        if n == 0:
+            # this should be eye() but it doesn't support
+            # the dtype kwarg yet so do it manually to save
+            # the copy required by eye(a.shape[0]).asdtype()
+            A = np.zeros(a.shape, dtype=np_dtype)
+            for k in range(a.shape[0]):
+                A[k, k] = 1.
+            return A
+
+        # note: to be consistent over contiguousness, C order is
+        # returned as that is what dot() produces and the most common
+        # paths through matrix_power will involve that. Therefore
+        # copies are made here to ensure the data ordering is
+        # correct for paths not going via dot().
+
+        if n < 0:
+            A = np.linalg.inv(a).copy()
```

Comment:
As discussed privately, this will depend on `dot()` support an integer version.  We will defer this problem.